File: backend/pkg/protocol/packet.go
package protocol

import "github.com/vmihailenco/msgpack/v5"

// Packet type constants — wire protocol over WebSocket binary frames.
const (
	TypeTerminalOutput = uint8(0x01)
	TypeTerminalInput  = uint8(0x02)
	TypeResize         = uint8(0x03)
	TypeSessionList    = uint8(0x10)
	TypeSessionStatus  = uint8(0x11)
	TypeScrollback     = uint8(0x12)
	TypeScrollbackDone = uint8(0x13)
	TypeAlert          = uint8(0x14)
	TypeKillSession    = uint8(0x15) // backend → agent: kill the session
	TypeRenameSession  = uint8(0x16) // agent → backend: rename the session
	TypeSpawnSession   = uint8(0x17) // backend → agent: spawn a new session
	TypeAuth           = uint8(0x20)
	TypeAuthOK         = uint8(0x21)
	TypeAuthFail       = uint8(0x22)
	TypeSubscribe      = uint8(0x30)
	TypeUnsubscribe    = uint8(0x31)
	TypeTmuxTree       = uint8(0x32) // agent → backend → clients: tmux pane hierarchy
	TypePing           = uint8(0xFF)
	TypePong           = uint8(0xFE)
)

// Packet is the wire format for all WebSocket messages.
type Packet struct {
	Type    uint8  `msgpack:"t"`
	Session string `msgpack:"s,omitempty"`
	Payload []byte `msgpack:"p,omitempty"`
}

// ResizePayload is used with TypeResize.
type ResizePayload struct {
	Cols uint16 `msgpack:"c"`
	Rows uint16 `msgpack:"r"`
}

// SessionStatusPayload is used with TypeSessionStatus.
type SessionStatusPayload struct {
	SessionID  string `msgpack:"id"`
	Status     string `msgpack:"status"` // active | exited | killed
	ExitCode   *int   `msgpack:"exit_code,omitempty"`
	Name       string `msgpack:"name,omitempty"`
	Command    string `msgpack:"cmd,omitempty"`
	TmuxBacked bool   `msgpack:"tmux_backed,omitempty"`
	TmuxTarget string `msgpack:"tmux_target,omitempty"`
}

// RenamePayload is used with TypeRenameSession.
type RenamePayload struct {
	SessionID string `msgpack:"id"`
	Name      string `msgpack:"name"`
}

// AlertPayload is used with TypeAlert.
type AlertPayload struct {
	Pattern string `msgpack:"p"`
	Excerpt []byte `msgpack:"e"`
}

// SpawnSessionPayload is used with TypeSpawnSession.
type SpawnSessionPayload struct {
	SessionID     string   `msgpack:"id"`
	Name          string   `msgpack:"name,omitempty"`
	Command       string   `msgpack:"cmd"`
	Cols          uint16   `msgpack:"cols"`
	Rows          uint16   `msgpack:"rows"`
	AlertPatterns []string `msgpack:"patterns,omitempty"`
	UseTmux       bool     `msgpack:"use_tmux,omitempty"`
	TmuxSplit     bool     `msgpack:"tmux_split,omitempty"` // split pane instead of new window
}

// TmuxTreePayload is used with TypeTmuxTree.
// Sent by the agent whenever the tmux topology changes.
type TmuxTreePayload struct {
	DeviceID string           `msgpack:"device_id"`
	Sessions []TmuxSessionTree `msgpack:"sessions"`
}

// TmuxSessionTree is one tmux session with its windows.
type TmuxSessionTree struct {
	Name    string           `msgpack:"name"`
	Windows []TmuxWindowTree `msgpack:"windows"`
}

// TmuxWindowTree is one tmux window with its panes.
type TmuxWindowTree struct {
	Index int            `msgpack:"index"`
	Name  string         `msgpack:"name"`
	Panes []TmuxPaneTree `msgpack:"panes"`
}

// TmuxPaneTree is one tmux pane mapped to a ccmux session ID.
type TmuxPaneTree struct {
	Index    int    `msgpack:"index"`
	CcmuxID  string `msgpack:"id"`    // ccmux session UUID for this pane
	Title    string `msgpack:"title"`
	Active   bool   `msgpack:"active"` // currently focused pane in the window
}

// SubscribePayload is used with TypeSubscribe.
type SubscribePayload struct {
	SessionID  string `msgpack:"id"`
	FromOffset string `msgpack:"offset,omitempty"` // Redis Streams offset
}

// Encode serialises a Packet to MessagePack bytes.
func (pkt *Packet) Encode() ([]byte, error) {
	return msgpack.Marshal(pkt)
}

// Decode deserialises MessagePack bytes into a Packet.
func Decode(data []byte) (*Packet, error) {
	var pkt Packet
	if err := msgpack.Unmarshal(data, &pkt); err != nil {
		return nil, err
	}
	return &pkt, nil
}

File: mobile/lib/core/protocol/packet.dart
import 'dart:typed_data';
import 'package:messagepack/messagepack.dart';

// Packet type constants — must exactly match backend/pkg/protocol/packet.go.
const int typeTerminalOutput = 0x01;
const int typeTerminalInput = 0x02;
const int typeResize = 0x03;
const int typeSessionList = 0x10;
const int typeSessionStatus = 0x11;
const int typeScrollback = 0x12;
const int typeScrollbackDone = 0x13;
const int typeAlert = 0x14;
const int typeAuth = 0x20;
const int typeAuthOK = 0x21;
const int typeAuthFail = 0x22;
const int typeSubscribe = 0x30;
const int typeUnsubscribe = 0x31;
const int typeTmuxTree = 0x32; // agent → backend → clients: tmux pane hierarchy
const int typePing = 0xFF;
const int typePong = 0xFE;

/// Wire packet matching the Go [Packet] struct with msgpack tags t/s/p.
class Packet {
  final int type;
  final String? session;
  final Uint8List? payload;

  const Packet({required this.type, this.session, this.payload});

  /// Encode to MessagePack binary, omitting null fields.
  Uint8List encode() {
    int fieldCount = 1; // always "t"
    if (session != null && session!.isNotEmpty) fieldCount++;
    if (payload != null && payload!.isNotEmpty) fieldCount++;

    final p = Packer();
    p.packMapLength(fieldCount);

    p.packString('t');
    p.packInt(type);

    if (session != null && session!.isNotEmpty) {
      p.packString('s');
      p.packString(session!);
    }
    if (payload != null && payload!.isNotEmpty) {
      p.packString('p');
      p.packBinary(payload!);
    }

    return p.takeBytes();
  }

  /// Decode a MessagePack binary frame into a Packet.
  static Packet decode(Uint8List data) {
    final u = Unpacker.fromList(data);
    final int mapLen = u.unpackMapLength();

    int type = 0;
    String? session;
    Uint8List? payload;

    for (int i = 0; i < mapLen; i++) {
      final key = u.unpackString()!;
      switch (key) {
        case 't':
          type = u.unpackInt()!;
        case 's':
          session = u.unpackString();
        case 'p':
          final binary = u.unpackBinary();
          payload = Uint8List.fromList(binary);
        default:
          u.unpackString(); // consume unknown fields
      }
    }
    return Packet(type: type, session: session, payload: payload);
  }
}

/// Encode a [ResizePayload] — matches protocol.ResizePayload{Cols, Rows}.
Uint8List encodeResizePayload(int cols, int rows) {
  final p = Packer();
  p.packMapLength(2);
  p.packString('c');
  p.packInt(cols);
  p.packString('r');
  p.packInt(rows);
  return p.takeBytes();
}

/// Encode a [SubscribePayload] — matches protocol.SubscribePayload{SessionID, FromOffset}.
Uint8List encodeSubscribePayload(String sessionId, {String fromOffset = ''}) {
  final int fieldCount = fromOffset.isNotEmpty ? 2 : 1;
  final p = Packer();
  p.packMapLength(fieldCount);
  p.packString('id');
  p.packString(sessionId);
  if (fromOffset.isNotEmpty) {
    p.packString('offset');
    p.packString(fromOffset);
  }
  return p.takeBytes();
}
